### PR TITLE
(PA-1086) Ship custom Windows rbconfig.rb

### PR DESCRIPTION
**Pull request by jcoconnor**
Use the method already in place for Solaris/AIX/Cross-Compiled-Linux to
ship a custom rbconfig.rb file that sets build_os to mingw32 instead of
cygwin.

*** WIP *** - still need to fill the contents of the custom rbconfig.rb file for each platform but putting PR up early to verify approach.

### CODE REVIEW

**Comment by Iristyle** [comment]
debaca5 seems like the right general direction

**Comment by ScottGarman** [comment]
Seems like a way to validate this would be to build an agent with this, and then try to install a gem with extensions (e.g, nokogiri) using our vendored gem command.

**Comment by McdonaldSeanp** [comment]
Agreed with @ScottGarman. @jcoconnor if you could try to gem install with the new packages and paste the output in a gist or something, that would be awesome

**Comment by jcoconnor** [comment]
Output from ```gem install nokogiri``` for Ruby 2.3.1/x86
```
C:\Windows\system32>cd "C:\Program Files (x86)\Puppet Labs\Puppet\bin"

C:\Program Files (x86)\Puppet Labs\Puppet\bin>environment.bat

C:\Program Files (x86)\Puppet Labs\Puppet\bin>gem install nokogiri
Fetching: mini_portile2-2.1.0.gem (100%)
Successfully installed mini_portile2-2.1.0
Fetching: nokogiri-1.7.1-x86-mingw32.gem (100%)
Nokogiri is built with the packaged libraries: libxml2-2.9.4, libxslt-1.1.29, zlib-1.2.11, libiconv-1.14.
Successfully installed nokogiri-1.7.1-x86-mingw32
Parsing documentation for mini_portile2-2.1.0
Installing ri documentation for mini_portile2-2.1.0
Parsing documentation for nokogiri-1.7.1-x86-mingw32
Installing ri documentation for nokogiri-1.7.1-x86-mingw32
Done installing documentation for mini_portile2, nokogiri after 17 seconds
2 gems installed

C:\Program Files (x86)\Puppet Labs\Puppet\bin>
```

**Comment by jcoconnor** [comment]
As above for Ruby 2.3.1/x64 build:
```
C:\Windows\system32>
C:\Windows\system32>
C:\Windows\system32>
C:\Windows\system32>cd "C:\Program Files\Puppet Labs\Puppet\bin"

C:\Program Files\Puppet Labs\Puppet\bin>environment.bat
C:\Program Files\Puppet Labs\Puppet\bin>gem install nokogiri
Fetching: mini_portile2-2.1.0.gem (100%)
Successfully installed mini_portile2-2.1.0
Fetching: nokogiri-1.7.1-x64-mingw32.gem (100%)
Nokogiri is built with the packaged libraries: libxml2-2.9.4, libxslt-1.1.29, zlib-1.2.11, libiconv-1.14.
Successfully installed nokogiri-1.7.1-x64-mingw32
Parsing documentation for mini_portile2-2.1.0
Installing ri documentation for mini_portile2-2.1.0
Parsing documentation for nokogiri-1.7.1-x64-mingw32
Installing ri documentation for nokogiri-1.7.1-x64-mingw32
Done installing documentation for mini_portile2, nokogiri after 12 seconds
2 gems installed
C:\Program Files\Puppet Labs\Puppet\bin>
```

**Comment by McdonaldSeanp** [comment]
👍 

Test output looks good, ready to merge from my perspective

**Comment by jcoconnor** [comment]
Just waiting to get the Ruby 2.1.9 build finished and tested.

**Comment by jcoconnor** [comment]
Have also verified the build on ```stable``` with these changes cherry-picked in and Ruby 2.1.9
Output for installing ```nokogiri```
```
C:\Program Files\Puppet Labs\Puppet\bin>environment.bat
C:\Program Files\Puppet Labs\Puppet\bin>gem install nokogiri
Fetching: mini_portile2-2.1.0.gem (100%)
Successfully installed mini_portile2-2.1.0
Fetching: nokogiri-1.7.1-x64-mingw32.gem (100%)
Nokogiri is built with the packaged libraries: libxml2-2.9.4, libxslt-1.1.29, zlib-1.2.11, libiconv-1.14.
Successfully installed nokogiri-1.7.1-x64-mingw32
Parsing documentation for mini_portile2-2.1.0
Installing ri documentation for mini_portile2-2.1.0
Parsing documentation for nokogiri-1.7.1-x64-mingw32
Installing ri documentation for nokogiri-1.7.1-x64-mingw32
Done installing documentation for mini_portile2, nokogiri after 10 seconds
2 gems installed
C:\Program Files\Puppet Labs\Puppet\bin>
```

**Comment by jcoconnor** [comment]
Have also verified that ```build_os="mingw32"``` on all four installation instances.

**Comment by ScottGarman** [comment]
:+1:

**Comment by Iristyle** [comment]
A couple of comments on this one... it doesn't look like the validation performed was necessarily correct for what we needed to test -- namely that a gem requiring native code compilation can use the DevKit, and furthermore, that it works in a non-Cygwin environment (ideally on a machine without Cygwin installed).

I still believe the contents of this PR are likely correct, but we should probably make sure the end result checks out.

/cc @scotje

File: resources/files/ruby_219/rbconfig/rbconfig-x86_64-w64-mingw32.rb [hunk 1]
```diff
@@ -0,0 +1,270 @@
+# Custom rbconfig.rb file obtained from Puppet 1.9.3 installation and modified
```
Comment (Iristyle) [inline resources/files/ruby_219/rbconfig/rbconfig-x86_64-w64-mingw32.rb]: I think you meant Puppet 4.9.3 / Ruby 2.1.9

File: resources/files/ruby_219/rbconfig/rbconfig-x86_64-w64-mingw32.rb [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+  CONFIG["host_vendor"] = "w64"
+  CONFIG["host_cpu"] = "x86_64"
+  CONFIG["host"] = "x86_64-w64-mingw32"
+  CONFIG["RUBY_VERSION_NAME"] = "$(RUBY_BASE_NAME)-$(ruby_version)"
+  CONFIG["RUBYW_BASE_NAME"] = "rubyw"
+  CONFIG["RUBY_BASE_NAME"] = "ruby"
+  CONFIG["build_os"] = "mingw32"
+  CONFIG["build_vendor"] = "unknown"
+  CONFIG["build_cpu"] = "x86_64"
+  CONFIG["build"] = "x86_64-unknown-cygwin"
```
Comment (Iristyle) [inline resources/files/ruby_219/rbconfig/rbconfig-x86_64-w64-mingw32.rb]: RubyInstaller (for an older pre-AIO agent based on Ruby 2.1.8) has:

```ruby
CONFIG["build_vendor"] = "pc"
CONFIG["build_cpu"] = "i686"
CONFIG["build"] = "i686-pc-mingw32"
```

**Comment by jcoconnor** [comment]
@Iristyle I have tried installing ```nokogiri``` on a non-cygwin system.
It installed ok.

**Comment by jcoconnor** [comment]
@Iristyle  output for installing ```pry``` gem:

```
C:\Program Files\Puppet Labs\Puppet\bin>gem install pry
Fetching: coderay-1.1.1.gem (100%)
Successfully installed coderay-1.1.1
Fetching: slop-3.6.0.gem (100%)
Successfully installed slop-3.6.0
Fetching: method_source-0.8.2.gem (100%)
Successfully installed method_source-0.8.2
Fetching: pry-0.10.4.gem (100%)
Successfully installed pry-0.10.4
invalid options: -SNw2
(invalid options are ignored)
Parsing documentation for coderay-1.1.1
Installing ri documentation for coderay-1.1.1
Parsing documentation for slop-3.6.0
Installing ri documentation for slop-3.6.0
Parsing documentation for method_source-0.8.2
Installing ri documentation for method_source-0.8.2
Parsing documentation for pry-0.10.4
Installing ri documentation for pry-0.10.4
Done installing documentation for coderay, slop, method_source, pry after 15 seconds
4 gems installed
C:\Program Files\Puppet Labs\Puppet\bin>
```

**Comment by jcoconnor** [comment]
However, attempting to build ```pry-byebug``` (using modified instructions from ) fails:
```
PS C:\tools\DevKit2\bin> gem install pry-byebug
Building native extensions.  This could take a while...
ERROR:  Error installing pry-byebug:
        ERROR: Failed to build gem native extension.

    current directory: C:/PROGRA~1/PUPPET~1/Puppet/sys/ruby/lib/ruby/gems/2.3.0/gems/byebug-9.0.6/ext/byebug
C:/PROGRA~1/PUPPET~1/Puppet/sys/ruby/bin/ruby.exe -r ./siteconf20170509-3120-noj8j0.rb extconf.rb
creating Makefile

current directory: C:/PROGRA~1/PUPPET~1/Puppet/sys/ruby/lib/ruby/gems/2.3.0/gems/byebug-9.0.6/ext/byebug
make "DESTDIR=" clean

current directory: C:/PROGRA~1/PUPPET~1/Puppet/sys/ruby/lib/ruby/gems/2.3.0/gems/byebug-9.0.6/ext/byebug
make "DESTDIR="
generating byebug-x64-mingw32.def
compiling breakpoint.c
/bin/sh: x86_64-w64-mingw32-gcc: command not found
make: *** [breakpoint.o] Error 127

make failed, exit code 2

Gem files will remain installed in C:/PROGRA~1/PUPPET~1/Puppet/sys/ruby/lib/ruby/gems/2.3.0/gems/byebug-9.0.6 for inspection.
Results logged to C:/PROGRA~1/PUPPET~1/Puppet/sys/ruby/lib/ruby/gems/2.3.0/extensions/x64-mingw32/2.3.0/byebug-9.0.6/gem_make.out
PS C:\tools\DevKit2\bin>

```

**Comment by jcoconnor** [comment]
@Iristyle Got the Ruby devout to install and initialise and was able to install/compile ```pry-byebug```
Output here:

```
PS C:\tools\DevKit2> ruby dk.rb init

Initialization complete! Please review and modify the auto-generated
'config.yml' file to ensure it contains the root directories to all
of the installed Rubies you want enhanced by the DevKit.
PS C:\tools\DevKit2> ruby dk.rb install
[ERROR] Unable to find RubyGems in site_ruby or core Ruby. Please
install RubyGems and rerun 'ruby dk.rb install'.
PS C:\tools\DevKit2>
PS C:\tools\DevKit2> ruby dk.rb install
[ERROR] Unable to find RubyGems in site_ruby or core Ruby. Please
install RubyGems and rerun 'ruby dk.rb install'.
PS C:\tools\DevKit2> ruby dk.rb install
[ERROR] Unable to find RubyGems in site_ruby or core Ruby. Please
install RubyGems and rerun 'ruby dk.rb install'.
PS C:\tools\DevKit2>
PS C:\tools\DevKit2>
PS C:\tools\DevKit2> ruby dk.rb install
[INFO] Installing 'C:/Program Files/Puppet Labs/Puppet/sys/ruby/lib/ruby/2.3.0/rubygems/defaults/operating_system.rb'
[INFO] Installing 'C:/Program Files/Puppet Labs/Puppet/sys/ruby/lib/ruby/site_ruby/devkit.rb'
PS C:\tools\DevKit2>
PS C:\tools\DevKit2>
PS C:\tools\DevKit2> gem install pry-byebug
Fetching: coderay-1.1.1.gem (100%)
Successfully installed coderay-1.1.1
Fetching: slop-3.6.0.gem (100%)
Successfully installed slop-3.6.0
Fetching: method_source-0.8.2.gem (100%)
Successfully installed method_source-0.8.2
Fetching: pry-0.10.4.gem (100%)
Successfully installed pry-0.10.4
Fetching: byebug-9.0.6.gem (100%)
Temporarily enhancing PATH to include DevKit...
Building native extensions.  This could take a while...
Successfully installed byebug-9.0.6
Fetching: pry-byebug-3.4.2.gem (100%)
Successfully installed pry-byebug-3.4.2
invalid options: -SNw2
(invalid options are ignored)
Parsing documentation for coderay-1.1.1
Installing ri documentation for coderay-1.1.1
Parsing documentation for slop-3.6.0
Installing ri documentation for slop-3.6.0
Parsing documentation for method_source-0.8.2
Installing ri documentation for method_source-0.8.2
Parsing documentation for pry-0.10.4
Installing ri documentation for pry-0.10.4
Parsing documentation for byebug-9.0.6
Installing ri documentation for byebug-9.0.6
Parsing documentation for pry-byebug-3.4.2
Installing ri documentation for pry-byebug-3.4.2
Done installing documentation for coderay, slop, method_source, pry, byebug, pry-byebug after 32 seconds
6 gems installed
PS C:\tools\DevKit2>
```

So - can we mark [PA-1086](https://tickets.puppetlabs.com/browse/PA-1086) as resolved now ? :1st_place_medal:

**Pull request merged**